test/runtests.jl

```diff
@@ -43,32 +43,62 @@ const url = "http://127.0.0.1:$port"
     @testset "Allocation profiling" begin
         done = Threads.Atomic{Bool}(false)
         # Schedule some work that's known to be expensive, to profile it
-        t = @async begin
+        workload() = @async begin
             for _ in 1:200
                 if done[] return end
                 global a = [[] for i in 1:1000]
                 yield()  # yield to allow the tests to run
             end
         end
 
-        req = HTTP.get("$url/allocs_profile?duration=3", retry=false, status_exception=false)
-        if !(isdefined(Profile, :Allocs) && isdefined(PProf, :Allocs))
-            # We should be tesing the Allocs profiling if we're on julia nightly.
-            @assert VERSION < v"1.8.0-DEV.1346"
-            @test req.status == 501  # not implemented
-        else
-            @test req.status == 200
-            @test length(req.body) > 0
+        @testset "allocs_profile endpoint" begin
+            done[] = false
+            t = workload()
+            req = HTTP.get("$url/allocs_profile?duration=3", retry=false, status_exception=false)
+            if !(isdefined(Profile, :Allocs) && isdefined(PProf, :Allocs))
+                @assert VERSION < v"1.8.0-DEV.1346"
+                @test req.status == 501  # not implemented
+            else
+                @test req.status == 200
+                @test length(req.body) > 0
+
+                data = read(IOBuffer(req.body), String)
+                # Test that there's something here
+                # TODO: actually parse the profile
+                @test length(data) > 100
+            end
+            @info "Finished `allocs_profile` tests, waiting for workload to finish."
+            done[] = true
+            wait(t)  # handle errors
+        end
 
-            data = read(IOBuffer(req.body), String)
-            # Test that there's something here
-            # TODO: actually parse the profile
-            @test length(data) > 100
+        @testset "allocs_profile_start/stop endpoints" begin
+            done[] = false
+            t = workload()
+            req = HTTP.get("$url/allocs_profile_start", retry=false, status_exception=false)
+            if !(isdefined(Profile, :Allocs) && isdefined(PProf, :Allocs))
+                @assert VERSION < v"1.8.0-DEV.1346"
+                @test req.status == 501  # not implemented
+            else
+                @test req.status == 200
+                @test String(req.body) == "Allocation profiling started."
+            end
 
+            req = HTTP.get("$url/allocs_profile_stop", retry=false, status_exception=false)
```

Comment:
i think the test is failing because you're not waiting long enough inbetween the start and stop. I'd probably add a `sleep(3)` here to mirror the `duration=3` in the `/allocs_profile?duration=3` test.

I don't love having tests that rely on sleep, since that's bound to be flaky, but it's probably good enough. 👍

Reply:
thanks! i was leaving this to debug in the morning, so nice to wake up to the answer!